sdks/java/core/src/main/java/org/apache/beam/sdk/coders/ZstdCoder.java

```diff
@@ -0,0 +1,110 @@
+import com.github.luben.zstd.ZstdCompressCtx;
+import com.github.luben.zstd.ZstdDecompressCtx;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/**
+ * Wraps an existing coder with Zstandard compression. It makes sense to use this coder when it's
+ * likely that the encoded value is quite large and compressible or when a dictionary is available
+ * to improve compression performance.
+ */
+public class ZstdCoder<T> extends StructuredCoder<T> {
+  private final Coder<T> innerCoder;
+  private final @Nullable byte[] dict;
+  private final int level;
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, byte[] dict, int level) {
+    return new ZstdCoder<>(innerCoder, dict, level);
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, byte[] dict) {
+    return new ZstdCoder<>(innerCoder, dict, Zstd.defaultCompressionLevel());
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, int level) {
+    return new ZstdCoder<>(innerCoder, null, level);
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder) {
+    return new ZstdCoder<>(innerCoder, null, Zstd.defaultCompressionLevel());
+  }
+
+  private ZstdCoder(Coder<T> innerCoder, @Nullable byte[] dict, int level) {
+    this.innerCoder = innerCoder;
+    this.dict = dict;
+    this.level = level;
+  }
+
+  @Override
+  public void encode(T value, OutputStream os) throws IOException {
+    ZstdCompressCtx ctx = new ZstdCompressCtx();
+    try {
+      ctx.setLevel(level);
+      ctx.setMagicless(true); // No magic since we know this will be compressed data on decode.
+      ctx.setDictID(false); // No dict ID since we initialize the coder with the expected dict.
+      ctx.loadDict(dict);
+
+      byte[] encoded = CoderUtils.encodeToByteArray(innerCoder, value);
+      ctx.setPledgedSrcSize(encoded.length);
+      byte[] compressed = ctx.compress(encoded);
+      ByteArrayCoder.of().encode(compressed, os);
+    } finally {
+      ctx.close();
+    }
+  }
+
+  @Override
+  public T decode(InputStream is) throws IOException {
+    ZstdDecompressCtx ctx = new ZstdDecompressCtx();
+    try {
+      ctx.setMagicless(true);
+      ctx.loadDict(dict);
+
+      byte[] compressed = ByteArrayCoder.of().decode(is);
+      int decompressedSize = (int) Zstd.decompressedSize(compressed, 0, compressed.length, true);
+      byte[] encoded = ctx.decompress(compressed, decompressedSize);
+      return CoderUtils.decodeFromByteArray(innerCoder, encoded);
+    } finally {
+      ctx.close();
+    }
+  }
+
```

Comment:
you have to implement equals since two zstd coders won't be equal if they use different parameters like dictionary

you should implement hashcode as well since your implementing equals

Reply:
Will do, this follows from wrongly subclassing StructuredCoder.

Reply:
Resolved in af74511.